### ISSUE TITLE
Lowest Common Ancestor 

### ISSUE THREAD

**Issue by jyoti246** [codezonediitj/pydatastructs#231]
#### Description of the problem
Using a naive approach we can find LCA between two nodes in a non cyclic graph in  ```O(n)``` time. But when there are ```O(n)``` queries i.e we need to get LCA of ```O(n)``` different pairs of nodes, the time complexity becomes ```O(n^2)```.
But by doing some pre processing we can reduce the complexity to ```O(nlogn)```. 
There are mainly two approaches:
Binary lifting 
using Segment tree and Eulars tour

Both of these can be implemented in pydatastructs/graphs/graph.py as classes

**Comment by czgdp1807** [comment]
Can you please provide some references(research paper, wikipedia, or university lectures) for the algorithm you are suggesting?

**Comment by jyoti246** [comment]
Here are they
For Binary Lifting:
https://cp-algorithms.com/graph/lca_binary_lifting.html
For Eulars Tour , Range Minimum query
https://www.geeksforgeeks.org/find-lca-in-binary-tree-using-rmq/

**Comment by czgdp1807** [comment]
> Let G be a tree

Quoting from [1]. This says that the graph should be a tree/DAG. The algorithm for binary trees is already available [here](https://github.com/codezonediitj/pydatastructs/blob/be07fdf45cde1a37f004b920cb2b4bc1e294c12b/pydatastructs/trees/binary_trees.py#L510), please see if any improvements can be made to it.

Is [2] relevant to the algorithm you are taking about?

References

.. [1] https://cp-algorithms.com/graph/lca_binary_lifting.html
.. [2] https://uhra.herts.ac.uk/bitstream/handle/2299/12152/906535.pdf;jsessionid=C46932A27B417CF192F2A9026D68706D?sequence=2

**Comment by czgdp1807** [comment]
Please follow the plan of action available [here](https://github.com/codezonediitj/pydatastructs/wiki/Plan-of-Action-for-the-Projects).

**Comment by robotjellyzone** [comment]
Hi @jyoti246 any update? are you still working on this ?

**Comment by Aakansha99** [comment]
Can I work on this issue?

**Comment by czgdp1807** [comment]
Read the stuff above and work accordingly.

**Comment by Rukmini-Meda** [comment]
I would like to work on this issue. Please assign this to me as part of GSSoC'20

**Comment by robotjellyzone** [comment]
Hi, @Rukmini-Meda yes, go ahead but please remember to follow the [plan of actions](https://github.com/codezonediitj/pydatastructs/wiki/Plan-of-Action-for-the-Projects) and please first discuss here regarding your solution and approach here.

**Comment by Rukmini-Meda** [comment]
Can you please assign this issue to me to avoid any confusion?

**Comment by robotjellyzone** [comment]
No we can't as there is no such rule of assigning here ! You can just straight away work on it without being assigned and if in case others also provide solution for this issue then the most appropriate and suitable solution will be merged !

**Comment by Rukmini-Meda** [comment]
Okay.

**Comment by robotjellyzone** [comment]
Any update @Rukmini-Meda? Are you still working on it!

**Comment by Rukmini-Meda** [comment]
Yes, I am working on it. Need little time. Will make a PR soon by tomorrow.

**Comment by robotjellyzone** [comment]
Hi @Rukmini-Meda are you still on it ?

**Comment by Rukmini-Meda** [comment]
I am sorry that I am not.

**Comment by Mintuagarwal** [comment]
Is it assigned? I have implemented binary-lifting in C++ for SPOJ problems. Can try it in python as well if not assigned.

**Comment by prshnt19** [comment]
Please read https://github.com/codezonediitj/pydatastructs/wiki/Issue-Policy and then you may proceed.

**Comment by subhangi2731** [comment]
can you please assign this issue to me @prshnt19

**Comment by Smit-create** [comment]
IMHO, since LCA  requires sparse table/segment tree for finding the minimum in a segment(Sparse table is prefered due to O(1) time for range min.queries), we should firstly focus on sparse table and segment tree additions and after that it only remains an implementation of LCA.

**Comment by czgdp1807** [comment]
> we should firstly focus on sparse table and segment tree additions and after that it only remains an implementation of LCA.

Sparse table can be added after discussion. I don't know if an issue is already open for it.

[Segment Tree](https://en.wikipedia.org/wiki/Segment_tree) is already there as far as I remember.

**Comment by Smit-create** [comment]
> Sparse table can be added after discussion.

https://github.com/codezonediitj/pydatastructs/issues/25 : Issue for Sparse Table. We should try adding this initially


> [Segment Tree](https://en.wikipedia.org/wiki/Segment_tree) is already there as far as I remember.

Yes, But can we make it customized for such operations: https://cp-algorithms.com/data_structures/segment_tree.html#toc-tgt-7, which using custom (`merge` function taken from user in form of `lambda` function) and builds on the given array of size `n`?. Specifically I am talking about something:
```py
>>> merge = lambda x, y: math.gcd(x, y)
>>> segTree = ODST(array, merge, dummy_variable) # array, custom merge function, dummy_variable (used while merging unwanted segment with required segment)
>>> segTree.build()
>>> segTree.query(1, 5) # Return gcd of elements array[1:6]
```

**Comment by rudrakshs-cerebras** [comment]
@czgdp1807 Working on a proposal for this implementation.
